### PR TITLE
corrected order of helm lint coalescing of multiple values files

### LINKED ISSUE #8880: Helm lint not pulling values outside of the default values file for schema validation
Hi 

Im not 100% if this is a bug or the intended functionality of the helm v3 lint command but I noticed an issue when I uplifted our ci pipeline from helm v2 to helm v3 (see actual versions below)

What is currently happening?
when I run `helm lint . --values=second_values.yaml` any value that is in second_values.yaml that is also in the values.yaml file of the chart will not be overridden with the values of second_values.yaml. 
Example
image

I was able to recreate the issue with a test chart (also attached as [testchart-0.1.0.zip](https://github.com/helm/helm/files/5369833/testchart-0.1.0.zip))

I dont know if this is a change in functionality. I am aware that the helm lint command handles schema checks differently between helm v2 and helm v3 however the schema used to validate the helm v3 lint command should be correct provided the second_values.yaml file is used.

Any help on this issue/question would be helpful as the only way we can work around this is by either disabling the helm lint command in our ci pipelines or removing some validation from our values.schema.json file

Output of `helm version`:
helm 3 installation
`helm version
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
`
helm 2 installation
`
helm2 version
Client: &version.Version{SemVer:"v2.16.5", GitCommit:"89bd14c1541fa93a09492010030fd3699ca65a97", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.16.1", GitCommit:"65cac393eeadea00384ea7108296ee6e5509c4de", GitTreeState:"clean"}
`

Output of `kubectl version`:
`kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:49Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:47:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}`

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  N/A

### PR DESCRIPTION
**Pull request by mluckam**
**What this PR does / why we need it**:

**Example**

The following command is run with the below value files and a json schema specifying service.port is required.

```bash
helm lint ./testchart --values ./testchart/second_values.yaml
```

_values.yaml_
```yaml
{
...
service:
  type: ClusterIP
  port:
...
}
```

_second_values.yaml_
```yaml
{
...
service:
  port: 80
...
}
```

The result is an error.
```bash
helm lint ./testchart --values ./testchart/second_values.yaml
==> Linting ./testchart
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - service: port is required


Error: 1 chart(s) linted, 1 chart(s) failed
```

See the attached chart below for the full, runnable example.
[testchart-0.1.0.zip](https://github.com/helm/helm/files/5850689/testchart-0.1.0.zip)

**Problem**

Running helm lint with the flag -f/--values specifies a user-provided values file.  When using helm lint if a user-provided values file exists, then the user-provided values file and chart provided values.yaml are coalesced to create a single values file.  When coalescing the values files if a value exists, see port above, in both the chart provided values.yaml and a user-provided values file, see second_values.yaml above, helm lint uses the value in the chart provided values.yaml.

This behavior is backwards as the user-provided values file should take precedent over the chart provided values.yaml.

**Solution**

The values files are coalesced in the following line of _pkg/lint/rules/values.go_
```go
values = chartutil.CoalesceTables(values, overrides)
```
The CoalesceTables function states the following
```go
// CoalesceTables merges a source map into a destination map.
//
// dest is considered authoritative.
func CoalesceTables(dst, src map[string]interface{}) map[string]interface{
```
Since dst is authoritative, overrides should be dst.  Also overrides is passed in so in order to not manipulate it, a deep copy of overrides should be used.

**Special notes for your reviewer**:
Fixes #8880 
Fixes #8892 

**If applicable**:
- [ ] this PR contains documentation
- [*] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Review by jdolitsky — APPROVED** [review]
this lgtm

File: pkg/lint/rules/values_test.go
Comment (bacongobbler) [inline pkg/lint/rules/values_test.go:107]: can you explain the reason behind this change to the existing unit tests? Could this potentially introduce breaking behaviour?

If at all possible can you please leave the existing test suite as it currently works? If this is intentional, can you please clarify why the values.yaml needs to contain a null value?

Comment (mluckam) [reply pkg/lint/rules/values_test.go:107]: The change to unit test TestValidateValuesFileSchemaOverrides was intentional.  TestValidateValuesFileSchemaOverrides was introduced and last modified by Issue #7984 which had the following goal:

“When support for helm lint --values was added, this support did not extend to the values file tests. This PR extends the support to values files, which lets the user schema-test the combination of the chart and the passed-in values.”

While TestValidateValuesFileSchemaOverrides does test the coalescing of a user-defined values file and the chart provided values.yaml, the unit test does not test overrides as its name specifies.  The field “password” given in the user-defined values file does not override a value in the chart provided values.yaml because the field “password” does not exist in the chart provided values.yaml.  Including the field “password” in the chart provided values.yaml verifies that an override of a field’s value has occurred with the user-provided value taking precedent.

Once the unit test was updated to take in an override, as described above, the overridden field “password” needed to be verified to ensure the correct value was used.  The function “validateValuesFile” coalesces the user-provided values file and the chart provided values.yaml, but does not return this coalesced object.  To ensure that the coalesced object has the correct value for “password” without being able to inspect the coalesced object, the field “password” in the chart provided values.yaml was given an invalid value of “null” while the user-provided values file was given a valid value of “swordfish”.  This ensured that if the function “validateValuesFile” coalesced correctly the unit test TestValidateValuesFileSchemaOverrides would pass and if the order of coalescing was incorrect that a failure would occur validating against the schema.

Comment (mluckam) [reply pkg/lint/rules/values_test.go:107]: The requested reasoning for the unit test change was provided above.  Can you provide feedback @bacongobbler  @jdolitsky as to how this pull request should proceed?

**Comment by mluckam** [comment]
The requested reasoning for the unit test change was provided above. Can you provide feedback @bacongobbler @jdolitsky as to how this pull request should proceed?

File: pkg/lint/rules/values_test.go
Comment (mluckam) [reply pkg/lint/rules/values_test.go:107]: @mattfarina are you able to provide feedback on this pull request?

Comment (bacongobbler) [reply pkg/lint/rules/values_test.go:107]: When we introduce new unit tests, we do not change the existing ones currently in place. Instead we add new tests to verify the new behaviour. That way we can ensure code that relied on prior behaviour still continues to work as expected. Changes like this one have introduced regressions in the past hence the concerns raised earlier.

Can you please keep the existing unit test in place and add a new unit test to verify the changes made above? A good practice to follow is described in https://golang.org/pkg/testing/, and there are examples in several areas of the helm code base, such as:

https://github.com/helm/helm/blob/6297c021cbda1483d8c08a8ec6f4a99e38be7302/pkg/action/lint_test.go#L79-L89

Comment (mluckam) [reply pkg/lint/rules/values_test.go]: Thank you for the feedback. The unit test in question was reverted and a new parameterized test was created as suggested.

**Comment by mluckam** [comment]
The circleci build failed on step test style with the following timeout error:
``` bash
GO111MODULE=on golangci-lint run
ERRO Timeout exceeded: try increasing it by passing --timeout option 
make: *** [Makefile:116: test-style] Error 4
```
I am able to run the following commands locally without error:
```bash
$ make test-style --debug
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating makefiles....
Updating goal targets....
 File 'test-style' does not exist.
Must remake target 'test-style'.
GO111MODULE=on golangci-lint run
Successfully remade target file 'test-style'.
```
```bash
$ golangci-lint run --verbose
INFO [config_reader] Config search paths: [./ /home/mluckam/development/helm/helm /home/mluckam/development/helm /home/mluckam/development /home/mluckam /home /]
INFO [config_reader] Used config file .golangci.yml
INFO [lintersdb] Active 14 linters: [deadcode dupl gofmt goimports golint gosimple govet ineffassign misspell nakedret staticcheck structcheck unused varcheck]
INFO [loader] Go packages loading at mode 575 (compiled_files|exports_file|types_sizes|deps|files|imports|name) took 6.1874238s
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 282.5927ms
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [linters context/goanalysis] analyzers took 0s with no stages
INFO [runner/skip dirs] Skipped 1 issues from dir internal/third_party/dep/fs by pattern (^|/)third_party($|/)
INFO [runner] Issues before processing: 43, after processing: 0
INFO [runner] Processors filtering stat (out/in): skip_dirs: 42/43, autogenerated_exclude: 42/42, cgo: 43/43, path_prettifier: 43/43, skip_files: 43/43, exclude-rules: 0/42, filename_unadjuster: 43/43, identifier_marker: 42/42, exclude: 42/42
INFO [runner] processing took 88.0427ms with stages: path_prettifier: 58.1572ms, autogenerated_exclude: 25.639ms, skip_dirs: 3.4317ms, identifier_marker: 586.7µs, exclude-rules: 207µs, cgo: 11.5µs, filename_unadjuster: 4.5µs, max_same_issues: 1.2µs, nolint: 1µs, exclude: 500ns, uniq_by_line: 500ns, diff: 400ns, skip_files: 300ns, max_per_file_from_linter: 200ns, path_shortener: 200ns, severity-rules: 200ns, max_from_linter: 200ns, source_code: 200ns, sort_results: 100ns, path_prefixer: 100ns
INFO [runner] linters took 384.675ms with stages: goanalysis_metalinter: 295.3923ms, unused: 1.1594ms
INFO File cache stats: 0 entries of total size 0B
INFO Memory: 70 samples, avg is 75.9MB, max is 140.3MB
INFO Execution took 6.88118s
```
Can you provide guidance on what would cause this issue?

@bacongobbler @jdolitsky

**Comment by bacongobbler** [comment]
Can you try rebasing your commits on top of master? There was a few fixes in master to address some CI issues.

**Review by bacongobbler — APPROVED** [review]
Thank you. LGTM! Once CodeQL passes we'll merge this.

**Comment by reitzig** [comment]
@bacongobbler Isn't this a bugfix? Any change we can get it sooner, i.e. with the next 3.5.x? :pray:

**Pull request merged**
